Broker does not close dependencies in correct order

With https://github.com/zeebe-io/zeebe/pull/2605 we wanted to fix a problem with the closing order. The problem was that the async director access a service which was not yet completely installed. 

I investigated this issue a bit further and it seems that on closing we also have the problem that the Partition (+StateSnapshotController) and Logblockindex is closed before the StreamProcessor, which is problematic. To reproduce this issue just run RejectIncompleteCommandsTest.

It seems that somehow the dependencies are not correctly handled? Please have a look at branch https://github.com/zeebe-io/zeebe/tree/2590-alternative-fix

**Be aware** That I added already a open and close future in the stream processor, which maybe causes this currently IDK.

I think this issue is not so urgent since this only happens if we close the broker during start up (when the stream processor is not completely opened).

```java

====
Broker startup time: 8188
====

10:57:55.982 [service-controller] [0.0.0.0:25602-zb-actors-1] ERROR io.zeebe.logstreams - start stream processor
10:57:55.982 [service-controller] [0.0.0.0:25602-zb-actors-1] ERROR io.zeebe.logstreams - OpenAsync
10:57:55.985 [zb-stream-processor] [0.0.0.0:25602-zb-actors-1] INFO  io.zeebe.logstreams - Recovering state of partition 1 from snapshot
10:57:56.002 [topology] [0.0.0.0:25602-zb-actors-0] DEBUG io.zeebe.broker.clustering - Updating partition information for partition 1 on Node{nodeId=0, commandApi=0.0.0.0:25602} with state LEADER
10:57:56.016 [] [atomix-cluster-events] DEBUG io.zeebe.gateway - Got membership event BrokerInfo{nodeId=0, partitionsCount=1, clusterSize=1, replicationFactor=1, addresses={command=0.0.0.0:25602}, partitionRoles={1=true}}
10:57:56.016 [] [atomix-cluster-events] DEBUG io.zeebe.broker.clustering - Member 0 received event ClusterMembershipEvent{type=METADATA_CHANGED, subject=Member{id=0, address=0.0.0.0:25603, properties={brokerInfo={"nodeId":0,"partitionsCount":1,"clusterSize":1,"replicationFactor":1,"addresses":{"command":"0.0.0.0:25602"},"partitionRoles":{"1":true}}}}, time=1560761876015}
10:57:56.018 [io.zeebe.gateway.impl.broker.cluster.BrokerTopologyManagerImpl] [0.0.0.0:25602-zb-actors-0] DEBUG io.zeebe.gateway - Received membership event: ClusterMembershipEvent{type=METADATA_CHANGED, subject=Member{id=0, address=0.0.0.0:25603, properties={brokerInfo={"nodeId":0,"partitionsCount":1,"clusterSize":1,"replicationFactor":1,"addresses":{"command":"0.0.0.0:25602"},"partitionRoles":{"1":true}}}}, time=1560761876015}
10:57:56.033 [exporter] [0.0.0.0:25602-zb-fs-workers-0] INFO  io.zeebe.broker.exporter - Recovering exporter 'exporter' from snapshot
10:57:56.117 [exporter] [0.0.0.0:25602-zb-fs-workers-0] INFO  io.zeebe.broker.exporter - Recovered exporter 'exporter' from snapshot at lastExportedPosition -1
10:57:56.145 [exporter] [0.0.0.0:25602-zb-fs-workers-0] INFO  io.zeebe.broker.exporter - Set event filter for exporters: ExporterEventFilter{acceptRecordTypes={NULL_VAL=true, COMMAND=true, EVENT=true, SBE_UNKNOWN=true, COMMAND_REJECTION=true}, acceptValueTypes={JOB_BATCH=true, DEPLOYMENT=true, WORKFLOW_INSTANCE_CREATION=true, NOOP=true, VARIABLE_DOCUMENT=true, WORKFLOW_INSTANCE_SUBSCRIPTION=true, ERROR=true, SBE_UNKNOWN=true, NULL_VAL=true, JOB=true, WORKFLOW_INSTANCE=true, VARIABLE=true, MESSAGE_START_EVENT_SUBSCRIPTION=true, TIMER=true, MESSAGE_SUBSCRIPTION=true, EXPORTER=true, INCIDENT=true, MESSAGE=true}}
10:57:56.190 [zb-stream-processor] [0.0.0.0:25602-zb-actors-1] INFO  io.zeebe.logstreams - Recovered state of partition 1 from snapshot at position -1
10:57:56.213 [] [main] INFO  io.zeebe.transport.endpoint - Registering endpoint for node '0' with address '0.0.0.0:25602' on transport 'gateway'
10:57:56.272 [commandApi.server-receiver] [0.0.0.0:25602-zb-actors-0] ERROR io.zeebe.broker.transport - Failed to deserialize message of type MESSAGE in client API
java.lang.RuntimeException: Could not deserialize object. Deserialization stuck at offset 40 of length 40
	at io.zeebe.msgpack.UnpackedObject.wrap(UnpackedObject.java:43) ~[classes/:?]
	at io.zeebe.broker.transport.commandapi.CommandApiMessageHandler.handleExecuteCommandRequest(CommandApiMessageHandler.java:127) ~[classes/:?]
	at io.zeebe.broker.transport.commandapi.CommandApiMessageHandler.onRequest(CommandApiMessageHandler.java:193) ~[classes/:?]
	at io.zeebe.transport.impl.ServerReceiveHandler.onFragment(ServerReceiveHandler.java:73) ~[classes/:?]
	at io.zeebe.transport.impl.TransportChannel.handleMessage(TransportChannel.java:153) ~[classes/:?]
	at io.zeebe.transport.impl.TransportChannel.receive(TransportChannel.java:129) ~[classes/:?]
	at io.zeebe.transport.impl.selector.ReadTransportPoller.processKey(ReadTransportPoller.java:110) ~[classes/:?]
	at org.agrona.nio.NioSelectedKeySet.forEach(NioSelectedKeySet.java:141) ~[agrona-1.0.2.jar:1.0.2]
	at io.zeebe.transport.impl.selector.ReadTransportPoller.processKeys(ReadTransportPoller.java:102) ~[classes/:?]
	at io.zeebe.transport.impl.selector.ReadTransportPoller.pollBlockingEnded(ReadTransportPoller.java:58) ~[classes/:?]
	at io.zeebe.util.sched.RunnableAdapter.lambda$wrapConsumer$2(RunnableAdapter.java:61) ~[classes/:?]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:90) [classes/:?]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:53) [classes/:?]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:189) [classes/:?]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:154) [classes/:?]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:135) [classes/:?]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:112) [classes/:?]
Caused by: java.lang.RuntimeException: Property 'name' has no valid value
	at io.zeebe.msgpack.value.ObjectValue.read(ObjectValue.java:128) ~[classes/:?]
	at io.zeebe.msgpack.UnpackedObject.wrap(UnpackedObject.java:39) ~[classes/:?]
	... 16 more
10:57:56.325 [commandApi.server-receiver] [0.0.0.0:25602-zb-actors-0] ERROR io.zeebe.broker.transport - Failed to deserialize message of type MESSAGE in client API
java.lang.RuntimeException: Could not deserialize object. Deserialization stuck at offset 35 of length 35
	at io.zeebe.msgpack.UnpackedObject.wrap(UnpackedObject.java:43) ~[classes/:?]
	at io.zeebe.broker.transport.commandapi.CommandApiMessageHandler.handleExecuteCommandRequest(CommandApiMessageHandler.java:127) ~[classes/:?]
	at io.zeebe.broker.transport.commandapi.CommandApiMessageHandler.onRequest(CommandApiMessageHandler.java:193) ~[classes/:?]
	at io.zeebe.transport.impl.ServerReceiveHandler.onFragment(ServerReceiveHandler.java:73) ~[classes/:?]
	at io.zeebe.transport.impl.TransportChannel.handleMessage(TransportChannel.java:153) ~[classes/:?]
	at io.zeebe.transport.impl.TransportChannel.receive(TransportChannel.java:129) ~[classes/:?]
	at io.zeebe.transport.impl.selector.ReadTransportPoller.processKey(ReadTransportPoller.java:110) ~[classes/:?]
	at org.agrona.nio.NioSelectedKeySet.forEach(NioSelectedKeySet.java:141) ~[agrona-1.0.2.jar:1.0.2]
	at io.zeebe.transport.impl.selector.ReadTransportPoller.processKeys(ReadTransportPoller.java:102) ~[classes/:?]
	at io.zeebe.transport.impl.selector.ReadTransportPoller.pollBlockingEnded(ReadTransportPoller.java:58) ~[classes/:?]
	at io.zeebe.util.sched.RunnableAdapter.lambda$wrapConsumer$2(RunnableAdapter.java:61) ~[classes/:?]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:90) [classes/:?]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:53) [classes/:?]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:189) [classes/:?]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:154) [classes/:?]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:135) [classes/:?]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:112) [classes/:?]
Caused by: java.lang.RuntimeException: Property 'correlationKey' has no valid value
	at io.zeebe.msgpack.value.ObjectValue.read(ObjectValue.java:128) ~[classes/:?]
	at io.zeebe.msgpack.UnpackedObject.wrap(UnpackedObject.java:39) ~[classes/:?]
	... 16 more
10:57:56.332 [commandApi.server-receiver] [0.0.0.0:25602-zb-actors-0] ERROR io.zeebe.broker.transport - Failed to deserialize message of type MESSAGE in client API
java.lang.RuntimeException: Could not deserialize object. Deserialization stuck at offset 46 of length 46
	at io.zeebe.msgpack.UnpackedObject.wrap(UnpackedObject.java:43) ~[classes/:?]
	at io.zeebe.broker.transport.commandapi.CommandApiMessageHandler.handleExecuteCommandRequest(CommandApiMessageHandler.java:127) ~[classes/:?]
	at io.zeebe.broker.transport.commandapi.CommandApiMessageHandler.onRequest(CommandApiMessageHandler.java:193) ~[classes/:?]
	at io.zeebe.transport.impl.ServerReceiveHandler.onFragment(ServerReceiveHandler.java:73) ~[classes/:?]
	at io.zeebe.transport.impl.TransportChannel.handleMessage(TransportChannel.java:153) ~[classes/:?]
	at io.zeebe.transport.impl.TransportChannel.receive(TransportChannel.java:129) ~[classes/:?]
	at io.zeebe.transport.impl.selector.ReadTransportPoller.processKey(ReadTransportPoller.java:110) ~[classes/:?]
	at org.agrona.nio.NioSelectedKeySet.forEach(NioSelectedKeySet.java:141) ~[agrona-1.0.2.jar:1.0.2]
	at io.zeebe.transport.impl.selector.ReadTransportPoller.processKeys(ReadTransportPoller.java:102) ~[classes/:?]
	at io.zeebe.transport.impl.selector.ReadTransportPoller.pollBlockingEnded(ReadTransportPoller.java:58) ~[classes/:?]
	at io.zeebe.util.sched.RunnableAdapter.lambda$wrapConsumer$2(RunnableAdapter.java:61) ~[classes/:?]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:90) [classes/:?]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:53) [classes/:?]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:189) [classes/:?]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:154) [classes/:?]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:135) [classes/:?]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:112) [classes/:?]
Caused by: java.lang.RuntimeException: Property 'timeToLive' has no valid value
	at io.zeebe.msgpack.value.ObjectValue.read(ObjectValue.java:128) ~[classes/:?]
	at io.zeebe.msgpack.UnpackedObject.wrap(UnpackedObject.java:39) ~[classes/:?]
	... 16 more
10:57:56.376 [] [main] INFO  io.zeebe.broker.test - Test execution time: 504
10:57:56.377 [] [main] INFO  io.zeebe.broker.system - Closing...
10:57:56.381 [exporter] [0.0.0.0:25602-zb-fs-workers-0] DEBUG io.zeebe.broker.exporter - Closed exporter director 'exporter'.
10:57:56.393 [] [zb-blocking-task-runner-3-0.0.0.0:25602] DEBUG io.zeebe.gateway - Closing gateway broker client ...
10:57:56.430 [] [zb-blocking-task-runner-3-0.0.0.0:25602] DEBUG io.zeebe.gateway - topology manager closed
10:57:56.430 [service-controller] [0.0.0.0:25602-zb-actors-0] ERROR io.zeebe.broker.clustering - Stop partition
10:57:56.431 [service-controller] [0.0.0.0:25602-zb-actors-0] DEBUG io.zeebe.broker.logstreams.restore.BrokerRestoreServer - Closed restore server for topics: log-replication-1, restore-info-1, snapshot-request-1, snapshot-info-request-1
10:57:56.462 [service-controller] [0.0.0.0:25602-zb-actors-0] DEBUG io.zeebe.logstreams.snapshot - Closed database from '/tmp/13b1c650-ff23-426c-96d0-f70d02460e24/data/partition-1/state/1_zb-stream-processor/runtime'.
10:57:56.464 [] [zb-blocking-task-runner-3-0.0.0.0:25602] DEBUG io.zeebe.gateway - transport closed
10:57:56.465 [] [zb-blocking-task-runner-3-0.0.0.0:25602] DEBUG io.zeebe.gateway - data frame receive buffer closed
10:57:56.465 [] [zb-blocking-task-runner-3-0.0.0.0:25602] DEBUG io.zeebe.gateway - Gateway broker client closed.
10:57:56.503 [zb-stream-processor] [0.0.0.0:25602-zb-actors-1] INFO  io.zeebe.processor - Start scanning the log for error events.
10:57:56.503 [zb-stream-processor] [0.0.0.0:25602-zb-actors-1] ERROR io.zeebe.util.actor - Actor failed in phase 'STARTED'. Continue with next job.
java.lang.AssertionError: null
	at org.rocksdb.Transaction.rollback(Transaction.java:215) ~[rocksdbjni-6.0.1.jar:?]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransaction.rollbackInternal(ZeebeTransaction.java:107) ~[classes/:?]
	at io.zeebe.db.impl.rocksdb.transaction.DefaultDbContext.runInNewTransaction(DefaultDbContext.java:174) ~[classes/:?]
	at io.zeebe.db.impl.rocksdb.transaction.DefaultDbContext.runInTransaction(DefaultDbContext.java:143) ~[classes/:?]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.ensureInOpenTransaction(ZeebeTransactionDb.java:154) ~[classes/:?]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.whileTrue(ZeebeTransactionDb.java:266) ~[classes/:?]
	at io.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.whileTrue(TransactionalColumnFamily.java:117) ~[classes/:?]
	at io.zeebe.logstreams.impl.log.index.LogBlockIndex.lookupBlockPosition(LogBlockIndex.java:114) ~[classes/:?]
	at io.zeebe.logstreams.impl.log.index.LogBlockIndex.lookupBlockAddress(LogBlockIndex.java:88) ~[classes/:?]
	at io.zeebe.logstreams.log.BufferedLogStreamReader.lookUpBlockAddressForPosition(BufferedLogStreamReader.java:341) ~[classes/:?]
	at io.zeebe.logstreams.log.BufferedLogStreamReader.seek(BufferedLogStreamReader.java:126) ~[classes/:?]
	at io.zeebe.logstreams.log.BufferedLogStreamReader.seekToFirstEvent(BufferedLogStreamReader.java:141) ~[classes/:?]
	at io.zeebe.logstreams.log.BufferedLogStreamReader.hasNext(BufferedLogStreamReader.java:194) ~[classes/:?]
	at io.zeebe.engine.processor.ReProcessingStateMachine.scanLog(ReProcessingStateMachine.java:173) ~[classes/:?]
	at io.zeebe.engine.processor.ReProcessingStateMachine.startRecover(ReProcessingStateMachine.java:154) ~[classes/:?]
	at io.zeebe.engine.processor.StreamProcessor.onActorStarted(StreamProcessor.java:165) ~[classes/:?]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:87) ~[classes/:?]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:53) [classes/:?]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:189) [classes/:?]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:154) [classes/:?]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:135) [classes/:?]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:112) [classes/:?]
10:57:56.515 [service-controller] [0.0.0.0:25602-zb-actors-0] DEBUG io.zeebe.logstreams.snapshot - Closed database from '/tmp/13b1c650-ff23-426c-96d0-f70d02460e24/data/partition-1/index/runtime'.
10:57:56.518 [zb-stream-processor] [0.0.0.0:25602-zb-actors-1] ERROR io.zeebe.util.actor - Actor failed in phase 'STARTED'. Continue with next job.
java.lang.AssertionError: null
	at org.rocksdb.Transaction.rollback(Transaction.java:215) ~[rocksdbjni-6.0.1.jar:?]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransaction.rollbackInternal(ZeebeTransaction.java:107) ~[classes/:?]
	at io.zeebe.db.impl.rocksdb.transaction.DefaultDbContext.runInNewTransaction(DefaultDbContext.java:174) ~[classes/:?]
	at io.zeebe.db.impl.rocksdb.transaction.DefaultDbContext.runInTransaction(DefaultDbContext.java:143) ~[classes/:?]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.ensureInOpenTransaction(ZeebeTransactionDb.java:154) ~[classes/:?]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.foreach(ZeebeTransactionDb.java:246) ~[classes/:?]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.foreach(ZeebeTransactionDb.java:232) ~[classes/:?]
	at io.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.forEach(TransactionalColumnFamily.java:99) ~[classes/:?]
	at io.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.forEach(TransactionalColumnFamily.java:95) ~[classes/:?]
	at io.zeebe.engine.state.deployment.DeploymentsState.foreachPending(DeploymentsState.java:73) ~[classes/:?]
	at io.zeebe.engine.processor.workflow.deployment.distribute.DeploymentDistributeProcessor.reprocessPendingDeployments(DeploymentDistributeProcessor.java:67) ~[classes/:?]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:87) ~[classes/:?]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:53) [classes/:?]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:189) [classes/:?]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:154) [classes/:?]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:135) [classes/:?]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:112) [classes/:?]
10:57:56.548 [] [raft-partition-group-system-0] INFO  io.atomix.core.impl.CorePrimitivesService - Stopped
10:57:56.579 [] [raft-server-system-partition-1] INFO  io.atomix.protocols.raft.impl.RaftContext - RaftServer{system-partition-1} - Transitioning to INACTIVE

```

--- a/broker-core/src/main/java/io/zeebe/broker/engine/EngineService.java
+++ b/broker-core/src/main/java/io/zeebe/broker/engine/EngineService.java
@@ -90,6 +90,7 @@ public class EngineService implements Service<EngineService> {
         .logStream(logStream)
         .actorScheduler(serviceContext.getScheduler())
         .additionalDependencies(partitionServiceName)
+        .additionalDependencies(serviceContext.getServiceName())
         .zeebeDb(partition.getZeebeDb())
         .serviceContainer(serviceContainer)
         .commandResponseWriter(new CommandResponseWriterImpl(commandApiTransport.getOutput()))

--- a/engine/src/main/java/io/zeebe/engine/processor/StreamProcessor.java
+++ b/engine/src/main/java/io/zeebe/engine/processor/StreamProcessor.java
@@ -69,6 +69,8 @@ public class StreamProcessor extends Actor implements Service<StreamProcessor> {
   private ProcessingStateMachine processingStateMachine;
 
   private Phase phase = Phase.REPROCESSING;
+  private CompletableActorFuture<Void> openFuture;
+  private CompletableActorFuture<Void> closeFuture = CompletableActorFuture.completed(null);
 
   protected StreamProcessor(final StreamProcessorBuilder context) {
     this.actorScheduler = context.getActorScheduler();
@@ -107,7 +109,7 @@ public class StreamProcessor extends Actor implements Service<StreamProcessor> {
 
   @Override
   public void start(ServiceStartContext startContext) {
-    startContext.async(openAsync());
+    startContext.async(openAsync(), true);
   }
 
   @Override
@@ -117,10 +119,10 @@ public class StreamProcessor extends Actor implements Service<StreamProcessor> {
 
   public ActorFuture<Void> openAsync() {
     if (isOpened.compareAndSet(false, true)) {
-      return actorScheduler.submitActor(this);
-    } else {
-      return CompletableActorFuture.completed(null);
+      openFuture = new CompletableActorFuture<>();
+      actorScheduler.submitActor(this);
     }
+    return openFuture;
   }
 
   @Override
@@ -133,12 +135,13 @@ public class StreamProcessor extends Actor implements Service<StreamProcessor> {
 
       lifecycleAwareListeners.forEach(l -> l.onOpen(processingContext));
     } catch (final Throwable e) {
-      onFailure();
+      onFailure(e);
       LangUtil.rethrowUnchecked(e);
     }
 
     try {
       processingStateMachine = new ProcessingStateMachine(processingContext, this::isOpened);
+      openFuture.complete(null);
 
       final ReProcessingStateMachine reProcessingStateMachine =
           new ReProcessingStateMachine(processingContext);
@@ -151,13 +154,13 @@ public class StreamProcessor extends Actor implements Service<StreamProcessor> {
           (v, throwable) -> {
             if (throwable != null) {
               LOG.error("Unexpected error on recovery happens.", throwable);
-              onFailure();
+              onFailure(throwable);
             } else {
               onRecovered();
             }
           });
     } catch (final RuntimeException e) {
-      onFailure();
+      onFailure(e);
       throw e;
     }
   }
@@ -214,10 +217,10 @@ public class StreamProcessor extends Actor implements Service<StreamProcessor> {
 
   public ActorFuture<Void> closeAsync() {
     if (isOpened.compareAndSet(true, false)) {
-      return actor.close();
-    } else {
-      return CompletableActorFuture.completed(null);
+      closeFuture = new CompletableActorFuture<>();
+      actor.close();
     }
+    return closeFuture;
   }
 
   @Override
@@ -239,11 +242,14 @@ public class StreamProcessor extends Actor implements Service<StreamProcessor> {
 
   @Override
   protected void onActorClosed() {
+    closeFuture.complete(null);
     LOG.debug("Closed stream processor controller {}.", getName());
   }
 
-  private void onFailure() {
+  private void onFailure(Throwable throwable) {
     phase = Phase.FAILED;
+    openFuture.completeExceptionally(throwable);
+    closeFuture = new CompletableActorFuture<>();
     isOpened.set(false);
     actor.close();
   }

--- a/qa/integration-tests/src/test/java/io/zeebe/broker/it/startup/BrokerReprocessingTest.java
+++ b/qa/integration-tests/src/test/java/io/zeebe/broker/it/startup/BrokerReprocessingTest.java
@@ -144,6 +144,16 @@ public class BrokerReprocessingTest {
   private Runnable restartAction = () -> {};
 
   @Test
+  public void shouldDirectlyRestart() {
+    // given
+
+    // when
+    reprocessingTrigger.accept(this);
+
+    // then - no error
+  }
+
+  @Test
   public void shouldCreateWorkflowInstanceAfterRestart() {
     // given
     deploy(WORKFLOW, "workflow.bpmn");

--- a/service-container/src/main/java/io/zeebe/servicecontainer/impl/ServiceContainerImpl.java
+++ b/service-container/src/main/java/io/zeebe/servicecontainer/impl/ServiceContainerImpl.java
@@ -27,9 +27,7 @@ import io.zeebe.util.sched.channel.ConcurrentQueueChannel;
 import io.zeebe.util.sched.future.ActorFuture;
 import io.zeebe.util.sched.future.CompletableActorFuture;
 import java.util.ArrayList;
-import java.util.HashMap;
 import java.util.List;
-import java.util.Map;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.TimeoutException;
@@ -52,10 +50,6 @@ public class ServiceContainerImpl extends Actor implements ServiceContainer {
   protected final ServiceDependencyResolver dependencyResolver = new ServiceDependencyResolver();
   protected final ConcurrentQueueChannel<ServiceEvent> channel =
       new ConcurrentQueueChannel<>(new ManyToOneConcurrentLinkedQueue<>());
-
-  protected final Map<ServiceName<?>, ServiceGroup> groups = new HashMap<>();
-
-  protected final Map<ServiceName<?>, List<ServiceController>> serviceListeners = new HashMap<>();
 
   protected final ActorScheduler actorScheduler;
 
@@ -93,6 +87,7 @@ public class ServiceContainerImpl extends Actor implements ServiceContainer {
   protected void onServiceEvent() {
     final ServiceEvent serviceEvent = channel.poll();
     if (serviceEvent != null) {
+      LOG.trace("{}", serviceEvent);
       dependencyResolver.onServiceEvent(serviceEvent);
     } else {
       actor.yield();
@@ -211,8 +206,7 @@ public class ServiceContainerImpl extends Actor implements ServiceContainer {
     try {
       containerCloseFuture.get(awaitTime, timeUnit);
     } catch (Exception ex) {
-      Loggers.SERVICE_CONTAINER_LOGGER.debug(
-          "Service container closing failed. Print dependencies.");
+      LOG.debug("Service container closing failed. Print dependencies.");
 
       final StringBuilder builder = new StringBuilder();
       dependencyResolver
@@ -227,7 +221,7 @@ public class ServiceContainerImpl extends Actor implements ServiceContainer {
                         });
               });
 
-      Loggers.SERVICE_CONTAINER_LOGGER.debug(builder.toString());
+      LOG.debug(builder.toString());
       throw ex;
     } finally {
       onClosed();
@@ -242,11 +236,18 @@ public class ServiceContainerImpl extends Actor implements ServiceContainer {
             state = ContainerState.CLOSING;
 
             final List<ActorFuture<Void>> serviceFutures = new ArrayList<>();
+
             dependencyResolver
-                .getControllers()
+                .getRootServices()
                 .forEach(
-                    (c) -> {
-                      serviceFutures.add(c.remove());
+                    c -> {
+                      final ActorFuture<Void> removeFuture = c.remove();
+                      actor.runOnCompletion(
+                          removeFuture,
+                          (v, t) -> {
+                            LOG.debug("Removed service {}", c.getServiceName());
+                          });
+                      serviceFutures.add(removeFuture);
                     });
 
             actor.runOnCompletion(

--- a/service-container/src/main/java/io/zeebe/servicecontainer/impl/ServiceController.java
+++ b/service-container/src/main/java/io/zeebe/servicecontainer/impl/ServiceController.java
@@ -45,7 +45,6 @@ import org.slf4j.Logger;
 @SuppressWarnings("rawtypes")
 public class ServiceController extends Actor {
   public static final Logger LOG = Loggers.SERVICE_CONTAINER_LOGGER;
-  public static final boolean IS_TRACE_ENABLED = LOG.isTraceEnabled();
 
   private final AwaitDependenciesStartedState awaitDependenciesStartedState =
       new AwaitDependenciesStartedState();
@@ -112,10 +111,6 @@ public class ServiceController extends Actor {
   private void onServiceEvent() {
     final ServiceEvent event = channel.poll();
     if (event != null) {
-      if (IS_TRACE_ENABLED) {
-        LOG.trace("Got {} in state {}", event, state.getClass().getSimpleName());
-      }
-
       state.accept(event);
     } else {
       actor.yield();
@@ -239,7 +234,7 @@ public class ServiceController extends Actor {
     }
 
     public void onStartFailed(Throwable t) {
-      LOG.error("Service failed to start while in AwaitStartState", t);
+      LOG.error("Service {} failed to start while in AwaitStartState", name, t);
       startFuture.completeExceptionally(t);
       state = awaitStopState;
       fireEvent(ServiceEventType.SERVICE_STOPPED);
@@ -289,9 +284,8 @@ public class ServiceController extends Actor {
       if (t.getType() == ServiceEventType.SERVICE_STOPPED) {
         injectors.values().stream().flatMap(Collection::stream).forEach(i -> i.uninject());
 
-        fireEvent(ServiceEventType.SERVICE_REMOVED);
-
         state = removedState;
+        fireEvent(ServiceEventType.SERVICE_REMOVED);
       }
     }
   }
@@ -337,7 +331,6 @@ public class ServiceController extends Actor {
     boolean isValid = true;
     boolean isAsync = false;
     boolean isInterruptible = false;
-    boolean stopOnCompletion = false;
     Runnable action;
 
     public void invalidate() {
@@ -394,10 +387,10 @@ public class ServiceController extends Actor {
       validCheck();
 
       if (!dependentServices.contains(name)) {
-        final Optional<ServiceController> contoller =
+        final Optional<ServiceController> controller =
             resolvedDependencies.stream().filter((c) -> c.name.equals(name)).findFirst();
 
-        if (!contoller.isPresent()) {
+        if (!controller.isPresent()) {
           final String errorMessage =
               String.format(
                   "Cannot remove service '%s' from context '%s'. Can only remove dependencies and services started through this context.",
@@ -540,19 +533,15 @@ public class ServiceController extends Actor {
     return String.format("%s in %s", name, state.getClass().getSimpleName());
   }
 
-  private void fireEvent(ServiceEventType evtType) {
+  public void fireEvent(ServiceEventType evtType) {
     fireEvent(evtType, null);
   }
 
-  private void fireEvent(ServiceEventType evtType, Object payload) {
+  public void fireEvent(ServiceEventType evtType, Object payload) {
     final ServiceEvent event = new ServiceEvent(evtType, this, payload);
 
     channel.add(event);
     container.getChannel().add(event);
-  }
-
-  public ConcurrentQueueChannel<ServiceEvent> getChannel() {
-    return channel;
   }
 
   public Set<ServiceName<?>> getDependencies() {

--- a/service-container/src/main/java/io/zeebe/servicecontainer/impl/ServiceDependencyResolver.java
+++ b/service-container/src/main/java/io/zeebe/servicecontainer/impl/ServiceDependencyResolver.java
@@ -131,12 +131,9 @@ public class ServiceDependencyResolver {
       final List<ServiceController> dependents = dependentServices.get(dependency);
 
       if (dependents != null) {
+        dependents.remove(controller);
         if (stoppingServices.contains(dependency)) {
-          boolean allStopped = true;
-          for (int i = 0; i < dependents.size() && allStopped; i++) {
-            allStopped &= !startedServices.contains(dependents.get(i).getServiceName());
-          }
-
+          final boolean allStopped = dependents.isEmpty();
           if (allStopped) {
             dependency.fireEvent(ServiceEventType.DEPENDENTS_STOPPED);
           }
@@ -148,8 +145,6 @@ public class ServiceDependencyResolver {
               controller,
               dependency.getServiceName());
         }
-
-        dependents.remove(controller);
       }
     }
 

--- a/service-container/src/main/java/io/zeebe/servicecontainer/impl/ServiceDependencyResolver.java
+++ b/service-container/src/main/java/io/zeebe/servicecontainer/impl/ServiceDependencyResolver.java
@@ -108,15 +108,11 @@ public class ServiceDependencyResolver {
 
     final List<ServiceController> dependents = this.dependentServices.get(controller);
     for (ServiceController dependentService : dependents) {
-      dependentService
-          .getChannel()
-          .add(new ServiceEvent(ServiceEventType.DEPENDENCIES_UNAVAILABLE, dependentService));
+      dependentService.fireEvent(ServiceEventType.DEPENDENCIES_UNAVAILABLE);
     }
 
     if (dependents.isEmpty()) {
-      controller
-          .getChannel()
-          .add(new ServiceEvent(ServiceEventType.DEPENDENTS_STOPPED, controller));
+      controller.fireEvent(ServiceEventType.DEPENDENTS_STOPPED);
     }
   }
 
@@ -132,28 +128,39 @@ public class ServiceDependencyResolver {
 
     final List<ServiceController> dependencies = resolvedDependencies.remove(controller);
     for (ServiceController dependency : dependencies) {
-      final List<ServiceController> deps = dependentServices.get(dependency);
+      final List<ServiceController> dependents = dependentServices.get(dependency);
 
-      if (deps != null) {
+      if (dependents != null) {
         if (stoppingServices.contains(dependency)) {
           boolean allStopped = true;
-          for (int i = 0; i < deps.size() && allStopped; i++) {
-            allStopped &= !startedServices.contains(deps.get(i));
+          for (int i = 0; i < dependents.size() && allStopped; i++) {
+            allStopped &= !startedServices.contains(dependents.get(i).getServiceName());
           }
 
           if (allStopped) {
-            dependency
-                .getChannel()
-                .add(new ServiceEvent(ServiceEventType.DEPENDENTS_STOPPED, dependency));
+            dependency.fireEvent(ServiceEventType.DEPENDENTS_STOPPED);
           }
+        } else {
+          // we should investigate this case - this seems to be wrong
+          // added this log to see if this ever happens
+          LOG.error(
+              "Removed service {}, but dependency service {} is not in closing yet, which means we do not remove dependents.",
+              controller,
+              dependency.getServiceName());
         }
 
-        deps.remove(controller);
+        dependents.remove(controller);
       }
     }
 
     installedServices.remove(controller.getServiceName());
-    dependentServices.remove(controller);
+
+    final List<ServiceController> dependents = dependentServices.remove(controller);
+    assert dependents == null || dependents.isEmpty()
+        : "Problem on dependency clean up, not closed dependents: "
+            + dependents
+            + " for controller "
+            + controller.getServiceName();
   }
 
   private void onServiceStarted(ServiceEvent event) {
@@ -220,6 +227,9 @@ public class ServiceDependencyResolver {
       }
     }
     this.resolvedDependencies.put(controller, resolvedDependencies);
+    if (dependencies.isEmpty()) {
+      rootServices.add(controller);
+    }
 
     /** resolve other services' dependencies which depend on this service */
     List<ServiceController> dependents = unresolvedDependencies.remove(controller.getServiceName());
@@ -233,6 +243,12 @@ public class ServiceDependencyResolver {
 
     this.dependentServices.put(controller, dependents);
     checkDependenciesAvailable(controller);
+  }
+
+  private final List<ServiceController> rootServices = new ArrayList<>();
+
+  public List<ServiceController> getRootServices() {
+    return rootServices;
   }
 
   private void checkDependenciesAvailable(ServiceController controller) {
@@ -251,13 +267,8 @@ public class ServiceDependencyResolver {
     }
 
     if (dependenciesAvailable) {
-      controller
-          .getChannel()
-          .add(
-              new ServiceEvent(
-                  ServiceEventType.DEPENDENCIES_AVAILABLE,
-                  controller,
-                  new ArrayList<>(resolvedDependencies)));
+      controller.fireEvent(
+          ServiceEventType.DEPENDENCIES_AVAILABLE, new ArrayList<>(resolvedDependencies));
     }
   }
 

--- a/service-container/src/test/java/io/zeebe/servicecontainer/impl/DependenciesTest.java
+++ b/service-container/src/test/java/io/zeebe/servicecontainer/impl/DependenciesTest.java
@@ -21,6 +21,7 @@ import static io.zeebe.servicecontainer.impl.ActorFutureAssertions.assertNotComp
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.Mockito.inOrder;
 import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
 import static org.mockito.Mockito.times;
 import static org.mockito.Mockito.verify;
 
@@ -30,6 +31,7 @@ import io.zeebe.servicecontainer.ServiceName;
 import io.zeebe.servicecontainer.ServiceStartContext;
 import io.zeebe.servicecontainer.ServiceStopContext;
 import io.zeebe.util.sched.future.ActorFuture;
+import io.zeebe.util.sched.future.CompletableActorFuture;
 import io.zeebe.util.sched.testing.ControlledActorSchedulerRule;
 import org.junit.Before;
 import org.junit.Rule;
@@ -212,6 +214,92 @@ public class DependenciesTest {
     // then
     final InOrder inOrder = inOrder(mockService1, mockService2, mockService3);
     inOrder.verify(mockService1, times(1)).stop(any(ServiceStopContext.class));
+    inOrder.verify(mockService2, times(1)).stop(any(ServiceStopContext.class));
+    inOrder.verify(mockService3, times(1)).stop(any(ServiceStopContext.class));
+  }
+
+  @Test
+  public void shouldStopDependentServicesFirstEvenWhenNotCompletelyStarted() {
+    final ActorFuture<Void> openFuture = new CompletableActorFuture<>();
+    final Service<Object> mockService1 =
+        spy(
+            new Service<Object>() {
+              @Override
+              public Object get() {
+                return null;
+              }
+
+              @Override
+              public void start(ServiceStartContext startContext) {
+                startContext.async(openFuture);
+              }
+            });
+    final Service<Void> mockService1b = mock(Service.class);
+    final Service<Object> mockService2 = mock(Service.class);
+    final Service<Object> mockService3 = mock(Service.class);
+
+    // given
+    final ServiceName<Void> service1b = ServiceName.newServiceName("service1b", Void.class);
+    serviceContainer.createService(service1, mockService1).dependency(service2).install();
+    serviceContainer.createService(service1b, mockService1b).dependency(service2).install();
+    serviceContainer.createService(service2, mockService2).dependency(service3).install();
+    serviceContainer.createService(service3, mockService3).install();
+    actorSchedulerRule.workUntilDone();
+
+    // when
+    serviceContainer.removeService(service3);
+    actorSchedulerRule.workUntilDone();
+
+    openFuture.complete(null);
+    actorSchedulerRule.workUntilDone();
+
+    // then
+    final InOrder inOrder = inOrder(mockService1, mockService1b, mockService2, mockService3);
+    inOrder.verify(mockService1b, times(1)).stop(any(ServiceStopContext.class));
+    inOrder.verify(mockService1, times(1)).stop(any());
+    inOrder.verify(mockService2, times(1)).stop(any(ServiceStopContext.class));
+    inOrder.verify(mockService3, times(1)).stop(any(ServiceStopContext.class));
+  }
+
+  @Test
+  public void shouldStopDependentServicesFirstEvenWhenNotCompletelyStartedOnClose() {
+    final ActorFuture<Void> openFuture = new CompletableActorFuture<>();
+    final Service<Object> mockService1 =
+        spy(
+            new Service<Object>() {
+              @Override
+              public Object get() {
+                return null;
+              }
+
+              @Override
+              public void start(ServiceStartContext startContext) {
+                startContext.async(openFuture);
+              }
+            });
+    final Service<Void> mockService1b = mock(Service.class);
+    final Service<Object> mockService2 = mock(Service.class);
+    final Service<Object> mockService3 = mock(Service.class);
+
+    // given
+    final ServiceName<Void> service1b = ServiceName.newServiceName("service1b", Void.class);
+    serviceContainer.createService(service3, mockService3).install();
+    serviceContainer.createService(service2, mockService2).dependency(service3).install();
+    serviceContainer.createService(service1b, mockService1b).dependency(service2).install();
+    serviceContainer.createService(service1, mockService1).dependency(service2).install();
+    actorSchedulerRule.workUntilDone();
+
+    // when
+    serviceContainer.closeAsync();
+    actorSchedulerRule.workUntilDone();
+
+    openFuture.complete(null);
+    actorSchedulerRule.workUntilDone();
+
+    // then
+    final InOrder inOrder = inOrder(mockService1, mockService1b, mockService2, mockService3);
+    inOrder.verify(mockService1b, times(1)).stop(any(ServiceStopContext.class));
+    inOrder.verify(mockService1, times(1)).stop(any());
     inOrder.verify(mockService2, times(1)).stop(any(ServiceStopContext.class));
     inOrder.verify(mockService3, times(1)).stop(any(ServiceStopContext.class));
   }

--- a/service-container/src/test/resources/log4j2-test.xml
+++ b/service-container/src/test/resources/log4j2-test.xml
@@ -8,7 +8,7 @@
   </Appenders>
 
   <Loggers>
-    <Logger name="io.zeebe" level="debug"/>
+    <Logger name="io.zeebe.servicecontainer" level="trace"/>
 
     <Root level="info">
       <AppenderRef ref="Console"/>

